### ISSUE TITLE
SOHO-6041: Show tooltips on disabled inputs and buttons.

### ISSUE THREAD

**Issue by clepore** [infor-design/enterprise#453]
**Summary**

When an {{input}} or {{button}} is disabled, it's mouse hover tooltip is not shown.

**Expected behavior**

As standard HTML does show title regardless of whether an input is disabled or enabled, I'd expect IDS to do the same. It is actually pretty useful to allow user to hover over a disabled field and give him feedback why is it disabled.

Just create an HTML page with this input:

```
<input type="text" disabled="disabled" title="I'm title"></input>
```

Without IDS, it shows the title. With IDS initialized, the title is removed at runtime (by the `Tooltip` component) and is not shown.

**Possible solution**

Enable tooltips to display on these elements whenever `settings.hover === true`, similar to what's been done for icons in #554

**Comment by darinw** [comment]
here is the workaround i have for angular

HTML
`<button *ngIf="buttonStates.view.tile.show" [disabled]="buttonStates.view.tile.disable" (click)="setViewType('Tile')" soho-button='icon' icon="grid" class="disabledTooltipHack">
                <div soho-tooltip title="Tiles View" placement="bottom"></div>
            </button>`

SCSS
`button.disabledTooltipHack {
  position: relative;

  span {
    position: initial !important;
    pointer-events: all !important;

    [soho-tooltip] {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      cursor: pointer;
      height: 100%;
      z-index: 2;
    }
  }
  svg {
    top: 4px !important;
  }
}`

**Comment by tmcconechy** [comment]
Wow nice that does work...We did not think of that.. Thanks!

```
<button class="btn-secondary" type="button" disabled><div title="test">Action</div></button>

<script>
  $('div[title]').tooltip();
</script>
```

**Comment by EdwardCoyle** [comment]
Some changes in #554 might help this along... may want to wait on doing any additional work on this til that PR is merged

**Comment by davidcarlsonberg** [comment]
Mouse events do not fire over `:disabled` elements (except on Firefox apparently). This precludes the application of a tooltip on a `:disabled` element. It would be possible to prevent the stripping of the `title` attribute on `disabled` elements and allowing the browser to handle but that would resolve in conflicting tooltips and titles on `:enabled` elements and the `title` on hover wouldn't adhere to IDS style specifications, unless we stripped the title only for `:enabled` and reapplied it for `:disabled` but just reading this sentence over gives me a headache.

It is suggested online that the `input` or `button`, if `:disabled`, be wrapped with another element that can fire events and thus tooltips (could also be a sibling element).

So, there is a thought to wrapping all `input` and `button` elements for this purpose; there may be other advantages to this. This would be a significant change and I wanted to voice it here in case I'm crazy or just overthinking the issue at hand.

Thoughts are welcome.

**Comment by tmcconechy** [comment]
I was thinking what we could do is limit disabled tooltips to buttons and skip inputs as thats a lesser used case.

I tested the workaround and it seemed to work ok.

```
<button class="btn-secondary" type="button" disabled><div title="test">Action</div></button>

<script>
  $('div[title]').tooltip();
</script>
```

So was thinking we could document that. If we want to take it further we could make the tooltip plugin correct this on the current buttons.

Or we could make the tooltip plugin
- test if the element is disabled
- when it is add an overlay over the element (opacity 0)
- the overlay has the eventing to show the tooltip
- disable/enable function would add/remove the overlay

I'd lean to the simplest solution of just adding an example of a working disabled button

**Comment by davidcarlsonberg** [comment]
Yea, the workaround works for buttons as they have content, as you mention. Since the ticket was geared toward addressing inputs, I assumed this was an essential component. I'm happy to work on a wrapping/overlay solution for both `button:disabled` and `input:disabled` in order to prevent two distinct workarounds and having the tooltip plugin handle both cases by addressing the `:disabled` nature instead of adding an additional div inside as that is inoperable on `input`.

Thanks for your input @tmcconechy !

@clepore Let me now if you have any thoughts when you get a moment as you introduced the issue.

**Comment by tmcconechy** [comment]
There was several request in Jira that were converted and merged and the summary of it all is both input and button where requested but over time maybe 10 people asked about buttons and 1 asked about inputs. (I'm not even sure what the use case would be on inputs).

So i think buttons would suffice and we could descope inputs

**Comment by EdwardCoyle** [comment]
@davidcarlsonberg -- scoped the input field-specific stuff to #579 for now.  I think it's worth exploring the generic wrapper element -- wouldn't be the first time it would've helped to have one.

**Comment by davidcarlsonberg** [comment]
Ok, proceeding with <button> workaround.

**Comment by brianjuan** [comment]
http://4100-beta0-enterprise.demo.design.infor.com/components/button/example-disabled-button-tooltip
Passed QA testing

**Comment by guoliang** [comment]
Is it possible a regression. ToolTip on disabled button doesn't work and if I used the above mentioned hack it looks like following then
![image](https://user-images.githubusercontent.com/542598/70317794-c451c480-181e-11ea-8c12-59df451af030.png)

This was tested on Chrome browser.

**Comment by tmcconechy** [comment]
@guoliang im not seeing any issues on this test page http://master-enterprise.demo.design.infor.com/components/button/example-disabled-button-tooltip 

That looks like a stock browser tooltip in Windows almost as if nothing is initialized in terms of the component.

**Comment by guoliang** [comment]
@tmcconechy weird. I tried to switch to the latest version as I was running 4.22.x but it was the same result. I tried to copy pasted the same template from the URL you gave as well. Regard initialization I think it should be right, since it works on the not disabled tooltip.

![disabled-tooltip-issue](https://user-images.githubusercontent.com/542598/70332966-e1988a00-1842-11ea-8675-0e9836be369f.gif)

**Comment by tmcconechy** [comment]
Im not sure. It does seem to work here in this example. Any difference in markup?
Or maybe a wrong element is getting the tooltip initialized against it?

**Comment by guoliang** [comment]
Now I see what's going on here.

Following proposed solution

```
<button id="test" class="btn-primary" type="button" title="btn-primary title/tooltip" disabled>
        <div class="disabled-tooltip" title="btn-primary title/tooltip"></div>
        Action
</button>
```
Is in fact two tooltip being initialized, one in the disabled and then one for the div element.
The first one being initialized by
`$('body').initialize({ locale: initialLocale });`
and then the second being init by the following line
`$('div[title]').tooltip();`

IMO, I think it the proposed solution is a bit hacky, and requires as mentioned two tooltip to be created for the case of disabled input/button elements.

**Comment by guoliang** [comment]
One case which the current solution going to fail is the following
`$('#test').data("tooltip").updated({ content: "hello world"})`
can be tested on the provided example page
http://master-enterprise.demo.design.infor.com/components/button/example-disabled-button-tooltip

I expect the disabled "Add Item" button to have a new tooltip text "hello world".

**Comment by tmcconechy** [comment]
We might be able to handle the case better by not invoking two tooltips. But we need the DIV its important for the trick to work or the disabled button will not get hover events.

If you want it fixed lets please make a new issue?

**Comment by guoliang** [comment]
@tmcconechy done
